### PR TITLE
feat: Add an existing SQS queue option to SNS event

### PR DESCRIPTION
**Pull request by 53ningen**
readable diff:
https://github.com/awslabs/serverless-application-model/pull/1231/files?w=1

*Issue #, if available:*
#854 

*Description of changes:*
Add an existing SQS queue option to SNS event

*Description of how you validated changes:*
- make pr
- verify transformed template deploys and application functions as expected

*Checklist:*

- [x] Write/update tests
- [x] `make pr` passes
- [x] Update documentation
- [x] Verify transformed template deploys and application functions as expected
- [x] Add/update example to `examples/2016-10-31`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by 53ningen** [comment]
ready for review

File: samtranslator/model/eventsources/push.py [hunk 1]
```diff
@@ -392,17 +392,42 @@ def to_cloudformation(self, **kwargs):
...
+            resources.append(queue_policy)
+            resources.append(subscription)
+            return resources
+
+        # SNS -> SQS(Existing) -> Lambda
         resources = []
-        queue = self._inject_sqs_queue()
-        queue_policy = self._inject_sqs_queue_policy(self.Topic, queue)
+        queue_arn = self.SqsSubscription.get('QueueArn', None)
+        queue_url = self.SqsSubscription.get('QueueUrl', None)
```
Comment (ShreyaGangishetty) [inline samtranslator/model/eventsources/push.py]: If `QueueArn` and `QueueUrl` are not given then we will be adding empty Strings for `Resource` and `Queues` fields in `PolicyDocument`. Instead we need to throw errors if these fields are not given.

**Comment by timoschilling** [comment]
The informations of `QueueArn` and `QueueUrl` seams to be redundant. Isn't it possible to build the `QueueUrl` from the `QueueArn`?

**Comment by chrisoverzero** [comment]
@timoschilling Gominingen described why this was sometimes not possible (and why a separate URL can be a feature) [back in the original issue](https://github.com/awslabs/serverless-application-model/issues/854#issuecomment-548891190). In short, `QueueArn` is rarely going to be a string. It will almost always be an object with a key of `Fn::GetAtt`.

I can imagine that and similar being features of the next round of the, uh, feature:

- If no `QueueUrl` specified:
  - If `QueueArn` is string, construct URL as string.
  - If it is `Fn::GetAtt` object, get logical ID, construct `Ref` object.
  - Bail on other intrinsics. 

…which sound like really nice checks, assuming I haven’t forgotten why they’re impossible.

File: samtranslator/model/eventsources/push.py (hunk 1 above)
Comment (53ningen) [reply samtranslator/model/eventsources/push.py]: Fixed in 40cba76

File: tests/model/eventsources/test_sns_event_source.py [hunk 2]
```diff
@@ -64,3 +64,29 @@ def test_to_cloudformation_passes_the_filter_policy(self):
...
         self.assertRaises(TypeError, self.sns_event_source.to_cloudformation)
+
+    def test_to_cloudformation_throws_when_queue_url_or_queue_arn_not_given(self):
+        sqsSubscription = {
+            'BatchSize': 5
+        }
+        self.sns_event_source.SqsSubscription = sqsSubscription
+        self.assertRaises(TypeError, self.sns_event_source.to_cloudformation)
+
+    def test_to_cloudformation_when_sqs_subscription_disable(self):
```
Comment (53ningen) [inline tests/model/eventsources/test_sns_event_source.py]: > Could you please add/update any of the existing unit tests for a condition where SqsSubscritption: False.

I've added on this line.

**Comment by 53ningen** [comment]
@ShreyaGangishetty 
Thank very much you for reviewing this PR and I'm sorry for my late response.
I have addressed your pull request comments.

Ready for review again.

**Comment by ljacobsson** [comment]
What's the status on this?

It looks like it was included in the v1.20.0 release, but I can't get it working and failing on [this line](https://github.com/aws/serverless-application-model/blob/bcb1aacc079bba29e16f9328105ede0551423625/samtranslator/model/eventsources/push.py#L427). I've even tried to define both `QueueArn` and `QueueUrl`

Also, there's no sign of this feature in the documentation.

**Comment by chrisoverzero** [comment]
>Also, there's no sign of this feature in the documentation.

It is in [the documentation](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-property-function-sns.html) and specification.

**Pull request merged**